### PR TITLE
Issue #249 Temporal design patterns implemented.

### PR DESCRIPTION
**Pull request by Jesonzavic**
Create temporal patterns for issue #249

- Adds Temporal Property, Snapshot, Effectivity, Audit Log, Temporal Object design patterns.

Pull request description

- Adds specified temporal patterns in #249

### CODE REVIEW

File: audit-log/src/main/java/com/iluwatar/auditlog/SimpleDate.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    SimpleDate that = (SimpleDate) o;
+    return date.equals(that.date);
+  }
```
Comment (iluwatar) [inline audit-log/src/main/java/com/iluwatar/auditlog/SimpleDate.java:61]: If you override equals I think you should also override HashCode. Check also if you could use https://projectlombok.org/features/EqualsAndHashCode

File: audit-log/src/main/java/com/iluwatar/auditlog/SimpleDate.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    if (setDate != null) {
+      return setDate;
+    }
+    return new SimpleDate(LocalDate.now(clock));
+  }
+
+  @Override
+  public String toString() {
+    return date.toString();
+  }
```
Comment (iluwatar) [inline audit-log/src/main/java/com/iluwatar/auditlog/SimpleDate.java:96]: See https://projectlombok.org/features/ToString

File: effectivity/README.md [hunk 3]
```diff
@@ -0,0 +1,158 @@
+---
```
Comment (iluwatar) [inline effectivity/README.md:1]: Please separate the rest of the patterns into their own pull requests. This is too big to review all at once.

**Review by iluwatar — CHANGES_REQUESTED** [review]

**Comment by iluwatar** [comment]
Additionally, please fix the Sonar analysis findings.

File: effectivity/README.md (hunk 3 above)
Comment (Jesonzavic) [reply effectivity/README.md:1]: No worries, Ill link them in the comments once I have the requests ready.

File: audit-log/README.md [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+* Is difficult to process, the tighter the accessing of temporal information is integrated, the 
+  less useful Audit Logs are.
+
+## Related patterns
+* [Temporal Property](https://martinfowler.com/eaaDev/TemporalProperty.html)
+* [Temporal Object](https://martinfowler.com/eaaDev/TemporalObject.html)
+* [Effectivity](https://martinfowler.com/eaaDev/Effectivity.html)
+
+## Credits
+* [Martin Fowler](https://martinfowler.com/eaaDev/AuditLog.html)
```
[the remark this answers is not in the document]
Comment (iluwatar) [reply audit-log/README.md:154]: Yeah this concerns only the directories, so the main directory of the pattern should be prefixed with temporal-.

**Pull request closed without merging**